File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l =1;
        int r = *max_element(piles.begin(),piles.end());
        while(l<r)
        {
            int k =l+(r-l)/2;
            if(can(piles , h , k))
            {
                r= k;
            }
            else
            {
                l=k+1;
            }
        }
        return l;
    }
    bool can(vector<int>piles, int h ,int k)
    {
        int hrs=0;
        for(int p :piles)
        {
            hrs+=ceil(double(p)/k);
        }
        return hrs <= h;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public static int minEatingSpeed(int[] piles, int h) {
        int hrs = 0;

        // int maxpile = Collections.max(Arrays.asList(piles));
        int max = Arrays.stream(piles).max().getAsInt();
        int mid = 0;
        int l = 1;
        int r = max;
        while (l < r) {
            mid = l + (r - l) / 2;
            int currhr = 0;
            for (int p : piles) {
                currhr += (Math.ceil((double)p / mid));
            }
            if (currhr <= h) {
                r = mid;
            } else {
                l = mid + 1;
            }

        }
        return l;
    }
}